### ISSUE TITLE
`get_predicted` error when `newdata` does not include all factor levels

### ISSUE THREAD

**Issue by vincentarelbundock** [easystats/insight#423]
In this model, `cyl` is a factor variable with 3 levels. I create a `nd` dataset to use for prediction, but this dataset only includes 2 rows -- thus excluding at least one of the factor levels from the prediction data. This dataset yields correct predictions with the standard `predict` function, but not with `get_predicted`:

``` r
library(insight)

mod <- lm(mpg ~ factor(cyl), data = mtcars)

nd <- structure(list(cyl = structure(2:3, .Label = c("4", "6", "8"), class = "factor")), out.attrs = list( dim = c(cyl = 2L), dimnames = list(cyl = c("cyl=6", "cyl=8"))), class = "data.frame", row.names = c(NA, -2L))

nd
#>   cyl
#> 1   6
#> 2   8

predict(mod, newdata = nd, se.fit = TRUE, interval = "confidence")
#> $fit
#>        fit      lwr      upr
#> 1 19.74286 17.25132 22.23439
#> 2 15.10000 13.33822 16.86178
#> 
#> $se.fit
#>         1         2 
#> 1.2182168 0.8614094 
#> 
#> $df
#> [1] 29
#> 
#> $residual.scale
#> [1] 3.223099

get_predicted(mod, newdata = nd)
#> Error in mm %*% vcovmat: non-conformable arguments


traceback()

5: get_predicted_se(x, predictions, data = data, ci_type = ci_type,
       vcov_estimation = vcov_estimation, vcov_type = vcov_type,
       vcov_args = vcov_args)
4: get_predicted_ci.default(x, predictions, data = args$data, ci_type = args$ci_type,
       ...)
3: get_predicted_ci(x, predictions, data = args$data, ci_type = args$ci_type,
       ...)
2: get_predicted.lm(mod, newdata = nd)
1: get_predicted(mod, newdata = nd)
```

**Comment by DominiqueMakowski** [comment]
Thanks @vincentarelbundock for testing and helping to improve this (relatively) new function, much appreciated :)

My hunch is that the error originates from somewhere around here:

https://github.com/easystats/insight/blob/1834558a7d702706981252e8e2cca6de9521c2c1/R/get_predicted_ci.R#L226-L237

I'm not sure how to easily fix it

**Comment by strengejacke** [comment]
The error occurs exactly here:

https://github.com/easystats/insight/blob/1834558a7d702706981252e8e2cca6de9521c2c1/R/get_predicted_ci.R#L294

The problem is that the variance-covariance matrix includes the intercept and the two levels from `cyl`:

``` r
library(insight)
mod <- lm(mpg ~ factor(cyl), data = mtcars)
get_varcov(mod)
#>              (Intercept) factor(cyl)6 factor(cyl)8
#> (Intercept)    0.9443969   -0.9443969   -0.9443969
#> factor(cyl)6  -0.9443969    2.4284491    0.9443969
#> factor(cyl)8  -0.9443969    0.9443969    1.6864230
```

while the model matrix, based on `newdata`, drops the lowest level and thus only includes intercept and one level of `cyl`:

``` r
nd <- structure(list(cyl = structure(2:3, .Label = c("4", "6", "8"), class = "factor")), out.attrs = list( dim = c(cyl = 2L), dimnames = list(cyl = c("cyl=6", "cyl=8"))), class = "data.frame", row.names = c(NA, -2L))
nd$mpg <- 0

model.matrix(mpg ~ factor(cyl), data = nd)
#>   (Intercept) factor(cyl)8
#> 1           1            0
#> 2           1            1
#> attr(,"assign")
#> [1] 0 1
#> attr(,"contrasts")
#> attr(,"contrasts")$`factor(cyl)`
#> [1] "contr.treatment"
```

This breaks the code where we matrix-multiply both:

``` r
library(insight)
mod <- lm(mpg ~ factor(cyl), data = mtcars)
vcovmat <- get_varcov(mod)

nd <- structure(list(cyl = structure(2:3, .Label = c("4", "6", "8"), class = "factor")), out.attrs = list( dim = c(cyl = 2L), dimnames = list(cyl = c("cyl=6", "cyl=8"))), class = "data.frame", row.names = c(NA, -2L))
nd$mpg <- 0

mm <- model.matrix(mpg ~ factor(cyl), data = nd)

# compute vcov for predictions
var_matrix <- mm %*% vcovmat %*% t(mm)
#> Error in mm %*% vcovmat: non-conformable arguments
```

I has this in _ggeffects_ as well and solved quite some time ago:

``` r
mod <- lm(mpg ~ factor(cyl), data = mtcars)
ggeffects::ggpredict(mod, "cyl [6,8]")
#> # Predicted values of mpg
#> 
#> cyl | Predicted |         95% CI
#> --------------------------------
#> 6   |     19.74 | [17.36, 22.13]
#> 8   |     15.10 | [13.41, 16.79]
```

Must look into the code how I did this. It was rather complicated reconstructing the original model matrix, or dropping the non-existing rows from the vcov, I think.

**Comment by vincentarelbundock** [comment]
One thing to consider is that `model.matrix` behaves differently based on whether we feed it a `terms` object or a `model` object:

``` r
mod <- lm(mpg ~ factor(cyl), data = mtcars)
nd <- structure(list(cyl = structure(2:3, .Label = c("4", "6", "8"), class = "factor")), out.attrs = list( dim = c(cyl = 2L), dimnames = list(cyl = c("cyl=6", "cyl=8"))), class = "data.frame", row.names = c(NA, -2L))
nd$mpg <- 30

model.matrix(mod, data = nd)
#>   (Intercept) factor(cyl)6 factor(cyl)8
#> 1           1            1            0
#> 2           1            0            1

model.matrix(terms(mod), data = nd)
#>   (Intercept) factor(cyl)8
#> 1           1            0
#> 2           1            1
```

One simple solution would be to use the model itself `x` instead of the terms. This is already what we do when the user does not supply a `newdata` argument, so it shouldn't pose any problems outside this function. See here:

https://github.com/easystats/insight/blob/master/R/get_predicted_ci.R#L224

I opened a PR https://github.com/easystats/insight/pull/427 to show how this could work; it solves my problem.

But there is some important-looking term manipulation going on for `hurdle` models to drop offsets. So either those offset terms would have to be dropped after creating `mm`, or `hurdle` will have the same problem identified above.

**Comment by strengejacke** [comment]
Can this be closed?

**Comment by vincentarelbundock** [comment]
Thanks for merging!

For the most part, this takes care of the problem. The only remaining issue is that we still use the old strategy for hurdle and zeroinfl.

Unfortunately, I don't really understand the code or it's purpose in that section.

Does anyone know if the term matching+dropping can be applied *after* to the model matrix, instead of *before* to the terms object? If so, we could use the same strategy for hurdle and zeroinfl models as well.

**Comment by strengejacke** [comment]
I'm not sure, but maybe this works for hurdle as well, when we use the model object?

``` r
library(pscl)
#> Classes and Methods for R developed in the
#> Political Science Computational Laboratory
#> Department of Political Science
#> Stanford University
#> Simon Jackman
#> hurdle and zeroinfl functions by Achim Zeileis
# Generate some zero-inflated data
set.seed(123)
N <- 100 # Samples
x <- runif(N, 0, 10) # Predictor
off <- rgamma(N, 3, 2) # Offset variable
yhat <- -1 + x * 0.5 + log(off) # Prediction on log scale
dat <- data.frame(y = NA, x, logOff = log(off)) # Storage dataframe

dat$y <- rpois(N, exp(yhat)) # Poisson process
dat$y <- ifelse(rbinom(N, 1, 0.3), 0, dat$y) # Zero-inflation process

# Fit zeroinfl model using 2 methods of offset input
m1 <- zeroinfl(y ~ offset(logOff) + x | 1, data = dat, dist = "poisson")
m2 <- zeroinfl(y ~ x | 1,
               data = dat,
               offset = logOff,
               dist = "poisson"
)

head(model.matrix(terms(m1), data = dat))
#> Error in offset(logOff): argument "offset" is missing, with no default
head(model.matrix(terms(m2), data = dat))
#>   (Intercept)        x
#> 1           1 2.875775
#> 2           1 7.883051
#> 3           1 4.089769
#> 4           1 8.830174
#> 5           1 9.404673
#> 6           1 0.455565
head(model.matrix(m1, data = dat))
#>   (Intercept)        x
#> 1           1 2.875775
#> 2           1 7.883051
#> 3           1 4.089769
#> 4           1 8.830174
#> 5           1 9.404673
#> 6           1 0.455565
```

<sup>Created on 2021-09-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by vincentarelbundock** [comment]
Interesting. But your example doesn't include a missing factor level, as in the original problem...

**Comment by strengejacke** [comment]
Oh, you're right. But seems to work anyway?

``` r
library(pscl)
# Generate some zero-inflated data
set.seed(123)
N <- 100 # Samples
x <- runif(N, 0, 10) # Predictor
off <- rgamma(N, 3, 2) # Offset variable
yhat <- -1 + x * 0.5 + log(off) # Prediction on log scale
f <- factor(sample(1:3, 100, TRUE), levels = 1:4)
dat <- data.frame(y = NA, x, logOff = log(off), f = f) # Storage dataframe

dat$y <- rpois(N, exp(yhat)) # Poisson process
dat$y <- ifelse(rbinom(N, 1, 0.3), 0, dat$y) # Zero-inflation process

# Fit zeroinfl model using 2 methods of offset input
m1 <- zeroinfl(y ~ offset(logOff) + x + f | 1, data = dat, dist = "poisson")
m2 <- zeroinfl(y ~ x + f| 1,
               data = dat,
               offset = logOff,
               dist = "poisson"
)

head(model.matrix(terms(m1), data = dat))
#> Error in offset(logOff): argument "offset" is missing, with no default
head(model.matrix(terms(m2), data = dat))
#>   (Intercept)        x f2 f3 f4
#> 1           1 2.875775  0  0  0
#> 2           1 7.883051  0  1  0
#> 3           1 4.089769  0  1  0
#> 4           1 8.830174  1  0  0
#> 5           1 9.404673  0  0  0
#> 6           1 0.455565  0  0  0
head(model.matrix(m1, data = dat))
#>   (Intercept)        x f2 f3
#> 1           1 2.875775  0  0
#> 2           1 7.883051  0  1
#> 3           1 4.089769  0  1
#> 4           1 8.830174  1  0
#> 5           1 9.404673  0  0
#> 6           1 0.455565  0  0
```

<sup>Created on 2021-09-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
wait, not for `m2`?

**Comment by vincentarelbundock** [comment]
Right, this is kind of tricky. 

That's why perhaps we want to apply the [same regexes as here](https://github.com/easystats/insight/blob/master/R/get_predicted_ci.R#L247) but to the colnames and rownames of the model matrix instead of the terms.

**Comment by vincentarelbundock** [comment]
@strengejacke wait! I think I found another example that fails with the current code. 

I'll investigate and report back.

**Comment by vincentarelbundock** [comment]
After some more investigation, I think I've concluded that this is a fundamental problem with `model.matrix`, and there isn't much that can be done.

One option would be to "pad" the `newdata` to include bad rows with the missing factors, and then drop it later in the code. But that feels like a terrible hack. Moreover, I think this is probably a very nice use-case that I ran into for very idiosyncratic reasons. So there's probably no need to introduce an ugly hack in `insight` just to fix this.

The code I pushed earlier may still help in some cases, and it doesn't break anything, so no need to rollback. Still, I'll close this issue now as I don't plan to work on this anymore, and because I don't feel there's a "real" issue here.

Thanks for your help with this!
